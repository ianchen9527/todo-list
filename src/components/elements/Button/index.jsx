import React from "react";
import styles from "./index.module.scss";
import PropTypes from "prop-types";
import cx from "classnames";

const getWrapperClassName = (disabled) =>
  cx({
    [styles["wrapper"]]: true,
    [styles["disabled"]]: disabled,
  });

const Button = ({ icon, loading, onClick, text, disabled }) => {
  return (
    <div className={getWrapperClassName(disabled)} onClick={onClick}>
      {icon && <div className={styles[`icon-${icon}`]} />}
      {text}
      {loading && <div className={styles["loading"]} />}
    </div>
  );
};

Button.propTypes = {
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  icon: PropTypes.string,
  text: PropTypes.string,
};

Button.defaultProps = {
  onClick: null,
  diabled: false,
  loading: false,
  icon: "add",
  text: "",
};

export default Button;
