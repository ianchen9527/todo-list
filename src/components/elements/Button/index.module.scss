@import "../../../stylesheets/constants/colors.scss";

.wrapper {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-content: center;
  height: 40px;
  background-color: $button;
  border: 1px solid $black;
  border-radius: 5px;
  cursor: pointer;
  color: $black;
  font-family: "Roboto";
  font-size: 14px;
  line-height: 21px;

  &:hover {
    background-color: $button-hover;
  }

  &.disabled {
    opacity: 0.5;
    cursor: default;

    &:hover {
      background-color: $button;
    }
  }

  @mixin icon {
    width: 20px;
    height: 20px;
    display: block;
    background-size: 20px 20px;
    margin-right: 15px;
  }

  .icon-add {
    @include icon();
    
    background-image: url("../../../assets/images/add.svg");
  }

  .icon-edit {
    @include icon();
    
    background-image: url("../../../assets/images/edit.svg");
  }

  .loading {
    width: 20px;
    height: 20px;
    display: block;
    background-size: 20px 20px;
    background-image: url("../../../assets/images/loading.gif");
    margin-left: 15px;
  }
}
