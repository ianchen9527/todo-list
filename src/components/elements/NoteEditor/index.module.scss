@import "../../../stylesheets/constants/colors.scss";

.wrapper {
  width: 70vw;
  height: 100%;
  overflow: auto;
  box-sizing: border-box;
  box-sizing: border-box;
  padding: 15px 20px;
    

  .title {
    border: 1px solid $black;
    padding: 15px;
    border-radius: 5px;
    box-sizing: border-box;
    margin-bottom: 15px;
    background-color: $button;
    color: $black;
    font-family: "Roboto";
    font-size: 16px;
    line-height: 24px;
    width: 100%;
  }

  .loading-view {
    width: 60px;
    height: 60px;
    display: block;
    background-size: 60px 60px;
    background-image: url("../../../assets/images/loading.gif");
  }

  .content {
    border: 1px solid $black;
    padding: 15px;
    border-radius: 5px;
    box-sizing: border-box;
    margin-bottom: 15px;
    background-color: $button;
    color: $black;
    font-family: "Roboto";
    font-size: 14px;
    line-height: 21px;
    height: 50vh;
    width: 100%;
  }

  .buttons-wrapper {
    width: 100%;
    display: flex;
    flex-flow: row wrap;
    justify-content: flex-start;
    align-items: center;

    .button-wrapper {
      width: 100px;
      margin-right: 15px;
    }
  }
}
