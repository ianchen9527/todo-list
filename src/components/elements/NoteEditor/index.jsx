import React from "react";
import styles from "./index.module.scss";
import PropTypes from "prop-types";
import Button from "../Button";

const renderLoadingView = () => <div className={styles["loading-view"]} />;

const renderEditor = ({
  note,
  onTitleChange,
  onContentChange,
  onCancelClick,
  onSaveClick,
  onDeleteClick,
}) =>
  note.id && (
    <>
      <input
        className={styles["title"]}
        value={note.title}
        onChange={(event) => onTitleChange(event.target.value)}
      />
      <textarea
        className={styles["content"]}
        onChange={(event) => onContentChange(event.target.value)}
        value={note.content}
      />
      <div className={styles["buttons-wrapper"]}>
        <div className={styles["button-wrapper"]}>
          <Button onClick={onCancelClick} text="Cancel" icon="cancel" />
        </div>
        <div className={styles["button-wrapper"]}>
          <Button onClick={onSaveClick} text="Save" icon="save" />
        </div>
        <div className={styles["button-wrapper"]}>
          <Button onClick={onDeleteClick} text="Delete" icon="delete" />
        </div>
      </div>
    </>
  );

const NoteEditor = (props) => {
  console.log(props);
  return (
    <div className={styles["wrapper"]}>
      {props.loading ? renderLoadingView() : renderEditor(props)}
    </div>
  );
};

NoteEditor.propTypes = {
  note: PropTypes.shape({ title: PropTypes.string, id: PropTypes.number }),
  loading: PropTypes.bool,
  onTitleChange: PropTypes.func,
  onContentChange: PropTypes.func,
  onCancelClick: PropTypes.func,
  onSaveClick: PropTypes.func,
  onDeleteClick: PropTypes.func,
};

NoteEditor.defaultProps = {
  note: {},
  loading: false,
  onTitleChange: null,
  onContentChange: null,
  onCancelClick: null,
  onSaveClick: null,
  onDeleteClick: null,
};

export default NoteEditor;
