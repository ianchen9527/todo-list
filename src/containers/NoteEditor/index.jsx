import { compose } from "redux";
import { connect } from "react-redux";
import {
  getNote,
  getLoading,
  getOnTitleChange,
  getOnContentChange,
} from "../../selectors/containers/noteEditor";
import NoteEditor from "../../components/elements/NoteEditor";
import withLifecycle from "@hocs/with-lifecycle";
import {
  getOnDidMount,
  getOnCancelClick,
  getOnSaveClick,
  getOnDeleteClick,
} from "../../sagas/noteEditor";
import { withRouter } from "react-router-dom";

const mapStateToProps = (state) => ({
  note: getNote(state),
  loading: getLoading(state),
});

const mapDispatchToProps = (dispatch, { history, match }) => ({
  onCancelClick: getOnCancelClick(dispatch, history),
  onSaveClick: getOnSaveClick(dispatch, history),
  onDeleteClick: getOnDeleteClick(dispatch, history),
  dispatch,
  match,
});

const mergeProps = (stateProps, dispatchProps) => ({
  ...stateProps,
  ...dispatchProps,
  onTitleChange: getOnTitleChange(dispatchProps.dispatch, stateProps.note),
  onContentChange: getOnContentChange(dispatchProps.dispatch, stateProps.note),
});

const lifeCycle = {
  onDidMount: getOnDidMount(),
};

export default compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps, mergeProps),
  withLifecycle(lifeCycle)
)(NoteEditor);
