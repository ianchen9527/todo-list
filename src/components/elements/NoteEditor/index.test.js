import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import NoteEditor from "./index";
import Button from "../Button";

describe("NoteEditor", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<NoteEditor />);
  });

  describe("default props", () => {
    it("should render without crashing", () => {
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe("props.loading", () => {
    let loading;
    describe("is true", () => {
      beforeEach(() => {
        loading = true;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });

    describe("is false", () => {
      beforeEach(() => {
        loading = false;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });

      describe("props.note", () => {
        let note;
        describe("is set", () => {
          beforeEach(() => {
            note = { id: 123, title: "title", content: "content" };
            wrapper.setProps({ note });
          });

          it("should render without crashing", () => {
            expect(toJSON(wrapper)).toMatchSnapshot();
          });

          describe("props.onTitleChange", () => {
            let onTitleChange, event;
            beforeEach(() => {
              onTitleChange = jest.fn();
              event = { target: { value: "value" } };
              wrapper.setProps({ onTitleChange });
            });

            afterEach(() => {
              onTitleChange.mockClear();
            });

            describe("when .title is changed", () => {
              beforeEach(() => {
                wrapper.find(".title").simulate("change", event);
              });

              it("should be triggered with event.target.value", () => {
                expect(onTitleChange).toHaveBeenCalledWith(event.target.value);
              });
            });
          });

          describe("props.onContentChange", () => {
            let onContentChange, event;
            beforeEach(() => {
              onContentChange = jest.fn();
              event = { target: { value: "value" } };
              wrapper.setProps({ onContentChange });
            });

            afterEach(() => {
              onContentChange.mockClear();
            });

            describe("when .content is changed", () => {
              beforeEach(() => {
                wrapper.find(".content").simulate("change", event);
              });

              it("should be triggered with event.target.value", () => {
                expect(onContentChange).toHaveBeenCalledWith(event.target.value);
              });
            });
          });

          describe("props.onCancelClick", () => {
            let onCancelClick;
            beforeEach(() => {
              onCancelClick = jest.fn();
              wrapper.setProps({ onCancelClick });
            });

            afterEach(() => {
              onCancelClick.mockClear();
            });

            describe("when the first Button is changed", () => {
              beforeEach(() => {
                wrapper.find(Button).at(0).simulate("click");
              });

              it("should be triggered with event.target.value", () => {
                expect(onCancelClick).toHaveBeenCalled();
              });
            });
          });

          describe("props.onSaveClick", () => {
            let onSaveClick;
            beforeEach(() => {
              onSaveClick = jest.fn();
              wrapper.setProps({ onSaveClick });
            });

            afterEach(() => {
              onSaveClick.mockClear();
            });

            describe("when the second Button is changed", () => {
              beforeEach(() => {
                wrapper.find(Button).at(1).simulate("click");
              });

              it("should be triggered with event.target.value", () => {
                expect(onSaveClick).toHaveBeenCalled();
              });
            });
          });

          describe("props.onDeleteClick", () => {
            let onDeleteClick;
            beforeEach(() => {
              onDeleteClick = jest.fn();
              wrapper.setProps({ onDeleteClick });
            });

            afterEach(() => {
              onDeleteClick.mockClear();
            });

            describe("when the third Button is changed", () => {
              beforeEach(() => {
                wrapper.find(Button).at(2).simulate("click");
              });

              it("should be triggered with event.target.value", () => {
                expect(onDeleteClick).toHaveBeenCalled();
              });
            });
          });
        });
      });
    });
  });
});
