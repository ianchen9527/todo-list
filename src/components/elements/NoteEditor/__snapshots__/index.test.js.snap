// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteEditor default props should render without crashing 1`] = `
<div
  className="wrapper"
/>
`;

exports[`NoteEditor props.loading is false props.note is set should render without crashing 1`] = `
<div
  className="wrapper"
>
  <input
    className="title"
    onChange={[Function]}
    value="title"
  />
  <textarea
    className="content"
    onChange={[Function]}
    value="content"
  />
  <div
    className="buttons-wrapper"
  >
    <div
      className="button-wrapper"
    >
      <Button
        disabled={false}
        icon="cancel"
        loading={false}
        onClick={null}
        text="Cancel"
      />
    </div>
    <div
      className="button-wrapper"
    >
      <Button
        disabled={false}
        icon="save"
        loading={false}
        onClick={null}
        text="Save"
      />
    </div>
    <div
      className="button-wrapper"
    >
      <Button
        disabled={false}
        icon="delete"
        loading={false}
        onClick={null}
        text="Delete"
      />
    </div>
  </div>
</div>
`;

exports[`NoteEditor props.loading is false should render without crashing 1`] = `
<div
  className="wrapper"
/>
`;

exports[`NoteEditor props.loading is true should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="loading-view"
  />
</div>
`;
